lib/ocpp/v201/charge_point.cpp

```diff
@@ -436,7 +456,33 @@ void ChargePoint::init_websocket() {
                           << " which is configurationSlot: " << configuration_slot;
             this->websocket_connection_status = WebsocketConnectionStatusEnum::Disconnected;
             this->message_queue->pause();
+            // B04.FR.01
+            // If offline period exceeds Offline threshold, send status notification for all all connectors.
+            int offline_threshold = this->device_model->get_value<int>(ControllerComponentVariables::OfflineThreshold);
+            if (offline_threshold > 0) {
+
+                // get the status of all the connectors
+                for (auto const& [evse_id, evse] : this->evses) {
+                    int number_of_connectors = evse->get_number_of_connectors();
+                    conn_states_per_evse_id_key conn_states_struct_key;
+                    conn_states_per_evse_id_value conn_states_struct_value;
+                    EVLOG_debug<< "LIBOCPP: evseId: " << evse_id << " numConn: " << number_of_connectors;
+                    for (int connector_id = 1; connector_id <= number_of_connectors; connector_id++) {
+                        conn_states_struct_key.evse_id = evse_id;
+                        conn_states_struct_key.connectior_id = connector_id;
+                        conn_state_per_evse[conn_states_struct_key].connector_status = evse->get_state(connector_id);
+                        EVLOG_debug << "conn_id: " << conn_states_struct_key.connectior_id
+                                   << " State: " << conn_state_per_evse[conn_states_struct_key].connector_status;
+                    }
+                }
 
+                this->offline_threshold_timer.interval(
```

Comment:
I still think it is a cleaner solution to just store the timestamp here and compare it when going back online. Now we have a bool variable and an entire timer object stored and the functionality is just the same.

`std::chrono::time_point<std::chrono::steady_clock>` should do the trick. It is a monotonic timer so no worries about it moving backwards or whatever.